### ISSUE TITLE
Need help - continuation keeps blocking

### ISSUE THREAD

**Issue by ghost** [Naios/fluent.cpp#5]
Hello,
I am running event loop in 2 threads to process incoming request. I have tried to use your library to avoid blocking the socket while getting data from database. However even with the continuation, the thread is blocked during the full execution. 
```cpp
auto handle_report(string &report_name, string &destination_directory, vector<json> &data)
{
	return cti::make_continuable<vector<vector<string>>, ExcelSchema>(
		[&report_name,
		 &destination_directory,
		 &data](auto &&promise) {
  vector<vector<string>> results;
ExcelSchema schema;
//I connect to Oracle and get data and also populate schema object
try{
promise.set_value(move(results), move(schema));
}
catch (exception &e)
			{
				promise.set_exception(std::current_exception());
			}
```
And then on my socket callback I write to a file and in last continuation once it is done I send reply back.
```cpp
handle_report(report_name, destination_directory, data).then([&download_folder, &dateformat](vector<vector<string>> results, ExcelSchema schema) {

			//write results to temporary xlsx file
			boost::filesystem::path file_name("/temp.xlsx");
			boost::filesystem::path report_file = download_folder / file_name;
			string dest_filename = report_file.string();

			lxw_workbook_options options = {.constant_memory = LXW_TRUE, .tmpdir = NULL};
			lxw_workbook *workbook = workbook_new_opt(dest_filename.c_str(), &options);
			lxw_worksheet *worksheet = workbook_add_worksheet(workbook, NULL);

			lxw_format *legend_format = workbook_add_format(workbook);
			format_set_bold(legend_format);
			format_set_align(legend_format, LXW_ALIGN_CENTER);
			lxw_format *date_format = workbook_add_format(workbook);
			format_set_num_format(date_format, "dd-mmm-yy");
			lxw_format *number_format = workbook_add_format(workbook);
			format_set_num_format(number_format, "0.00");
			schema.createLegend(results[0]);
			vector<string> read_types = schema.getReadTypes();
			for (uint32_t row = 0; row < results.size(); ++row)
			{
				for (uint32_t col = 0; col < results[row].size(); ++col)
				{
					string value = results[row][col];
					if (value != "NULL")
					{
						if (row == 0)
						{
							worksheet_write_string(worksheet, row, col, value.c_str(), legend_format);
						}
						else if (read_types[col] == "string")
						{
							worksheet_write_string(worksheet, row, col, value.c_str(), nullptr);
						}
						else if (read_types[col] == "number")
						{
							double value_num;
							istringstream ss(value);
							ss >> value_num;
							worksheet_write_number(worksheet, row, col, value_num, number_format);
						}
						else if (read_types[col] == "date")
						{
							Date date(value, dateformat);
							lxw_datetime datetime = {date.GetYear(), date.GetMonth(), date.GetDay(), 0, 0, 0.0};
							worksheet_write_datetime(worksheet, row, col, &datetime, date_format);
						}
						else
						{
							throw invalid_argument("Unexpected column type: " + read_types[col]);
						}
					}
				}
			}
			workbook_close(workbook);

			//load file into memory and delete temp file
			std::ifstream t(dest_filename.c_str());
			std::string temp_data((std::istreambuf_iterator<char>(t)),
								  std::istreambuf_iterator<char>());
			t.close();
			boost::filesystem::remove(dest_filename);

			return temp_data;
		})
		.then([&req](string data) {
			//send response
			static h2o_generator_t generator = {NULL, NULL};
			h2o_iovec_t body = h2o_strdup(&req->pool, data.c_str(), data.length());
			req->res.status = 200;
			req->res.reason = "OK";
			h2o_start_response(req, &generator);
			h2o_send(req, &body, 1, H2O_SEND_STATE_FINAL);
		})
		.fail([&req](std::exception_ptr err) {
			try
			{
				std::rethrow_exception(err);
			}
			catch (std::exception const &e)
			{
				json error;
				error["err"] = e.what();
				vector<uint8_t> msgpack_response = json::to_msgpack(error);
				string msgpack_string = string(msgpack_response.begin(), msgpack_response.end());

				static h2o_generator_t generator = {NULL, NULL};
				h2o_iovec_t body = h2o_strdup(&req->pool, msgpack_string.c_str(), msgpack_string.length());
				req->res.status = 500;
				req->res.reason = "ERROR";
				h2o_start_response(req, &generator);
				h2o_send(req, &body, 1, H2O_SEND_STATE_FINAL);
			}
		});
	}
```

**Comment by Naios** [comment]
From my point of understanding you intend that

> //I connect to Oracle and get data and also populate schema object

is non-blocking during the execution of the continuation chain.
This is only possible if your database API calls are asynchronous too, otherwise they will block the current thread by design.
In order to make this work you would need to execute your query asynchronously and call the promise once the query has finished.

For instance with a redis database it would like this ([cpp_redis](https://github.com/Cylix/cpp_redis)):

```cpp
auto redis_get(std::string query) {
  return cti::make_continuable<cpp_redis::reply>([&, query = std::move(query)] mutable {
    client.get(std::move(query), [promise = std::forward<decltype(promise)>(promise)]
      (cpp_redis::reply reply) {
        promise.set_value(reply);
      });
  });
}

// ...

redis_get("hello").then([](cpp_redis::reply) {
  // ...
});
```

I'm not familiar with the Oracle API, but probably they are offering similar capabilities.

By the fact you are using the library doesn't mean your API calls magically are converted into asynchronous calls. If you invoke the promises instantly one by one with blocking calls between, your continuation chain still will be synchronous.

Just a note, both calls to the promise shall never throw an exception:
> try{
promise.set_value(move(results), move(schema));
}
catch (exception &e)
			{
				promise.set_exception(std::current_exception());
			}

**Comment by ghost** [comment]
Thank you for your reply. I completely understand that the best way would be to run the calls to the DB asynchronously so that the thread can listen to another connection while processing the work from previous call. However with Oracle things are difficult,as there are only really low level calls in OCI library to do this and this would add so much complexity to the whole thing. So I have read something about similar libraries Folly/Futures and Seastar engine. I am not a big fan of a Facebook,their libraries never worked for me and I wasted hours of time to get it working. Seastar idea about futures and promises seemed interesting and at least there from tutorial it seemed  that when they called function.then() the loop continued without being blocked by that function. It may be that I have got a fake assumption it is all magically like that since they also used asynchronous calls inside the futures. So yeah, I am kind of a lost in this multi-threading stuff. Am I really wrong on this?

At least if I studied how Nginx works I understood the idea, that when there is a job that can't be done asynchronously, another worked thread is created to handle it so it doesn't block the event loop. So I thought your library does exactly that behind the hood.

**Comment by Naios** [comment]
The Seastar engine works through CPU bound threads which are running lights fibers on them.
Probably they are doing the same as decribed above since a waiting future will suspend the current fiber instead of the current thread. Basically this won't help you out with your particular issue.

Continuable completely isn't thread or execution engine bound and thus is just an abstraction over a callback chain.

In order to fix your particular issue I would suggest you the common way of running a separate pool of threads for blocking calls and use those for invoking such a blocking call. This isn't the most elegant way but it works in every case. Then you are still able to use the continuable library while not blocking your main thread. Otherwise you would need to use the non blocking database API calls as described above.

**Comment by Naios** [comment]
Ah `static h2o_generator_t generator = {NULL, NULL};` probably this is dangerous because this isn't threadsafe.

**Comment by ghost** [comment]
> The Seastar engine works through CPU bound threads which are running lights fibers on them.
Probably they are doing the same as decribed above since a waiting future will suspend the current fiber instead of the current thread. 

So if I understand correctly, so even with multiple fibers I would be still blocking the thread like I do now, right? This seems to make only sense if all the jobs run inside are async operations ?

> In order to fix your particular issue I would suggest you the common way of running a separate pool of threads for blocking calls and use those for invoking such a blocking call. This isn't the most elegant way but it works in every case.

Do I have to write it from the scratch or could you give some recommendation on existing libraries that already handle this? 

> Otherwise you would need to use the non blocking database API calls as described above.

So far I have tried Ocilib library, but maybe there is an asynchronous ODBC driver, I will investigate.

**Comment by Naios** [comment]
> So if I understand correctly, so even with multiple fibers I would be still blocking the thread like I do now, right? This seems to make only sense if all the jobs run inside are async operations ?

Yes I think so.

> Do I have to write it from the scratch or could you give some recommendation on existing libraries that already handle this?

You could use boost::asio for that

**Comment by ghost** [comment]
Okay and about your example
```
auto redis_get(std::string query) {
cout << "Async call started" << endl;
  return cti::make_continuable<cpp_redis::reply>([&, query = std::move(query)] mutable {
    client.get(std::move(query), [promise = std::forward<decltype(promise)>(promise)]
      (cpp_redis::reply reply) {
        promise.set_value(reply);
      });
  });
}

cout << "Starting loop" << endl;
redis_get("hello").then([](cpp_redis::reply) {
   //this should get called when the async call is finished, right?
   cout << "Async call finished" << endl;
});
//if the above call is async, the main loop will continue instantly here?
cout << "Ending loop" <<< endl;
```
Do I understand correctly that this will print
```
Starting loop
Async call started
Ending loop
Async call finished
Starting loop
....
```
So basically what your library does is to handle the callback of asynchronous call, so that the main loop will continue and the "then" clause gets executed after the async callback has finished?

**Comment by Naios** [comment]
Depends on the scheduling. The async call finished might be executed before ending loop

**Comment by Naios** [comment]
I'll close this since the original issue seems to be resolved.
If you have any further questions open a new ticket please.
